x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyMaintenanceService.java

```diff
@@ -145,44 +146,35 @@ void concludeMaintenance() {
-                            .filter(indexName -> shouldRemoveIndex(getIndexResponse, policies, indexName))
-                            .toArray(String[]::new);
-                        deleteIndices(removeIndices);
-                    } else {
-                        logger.debug("Skipping enrich index cleanup since enrich policy was executed while gathering indices");
-                        concludeMaintenance();
-                    }
-                }
-
-                @Override
-                public void onFailure(Exception e) {
-                    logger.error("Failed to get indices during enrich index maintenance task", e);
-                    concludeMaintenance();
-                }
-            });
-        } else {
-            concludeMaintenance();
-        }
+        // Get cluster state first because for a new index to exist there then it has to have already been added to the inflight list
+        // If the new index is in the cluster state but not in the inflight list, then that means that the index name was removed from the
+        // list inflight list because it has already been promoted or because the policy execution failed.
+        ClusterState clusterState = clusterService.state();
+        Set<String> inflightPolicyExecutionIndices = enrichPolicyLocks.inflightPolicyIndices();
+        final Map<String, EnrichPolicy> policies = EnrichStore.getPolicies(clusterState);
+        logger.debug(() -> "Working enrich indices excluded from maintenance [" + String.join(", ", inflightPolicyExecutionIndices) + "]");
+        String[] removeIndices = clusterState.metadata()
+            .indices()
+            .values()
+            .stream()
+            .filter(indexMetadata -> indexMetadata.getIndex().getName().startsWith(EnrichPolicy.ENRICH_INDEX_NAME_BASE))
+            .filter(indexMetadata -> shouldRemoveIndex(indexMetadata, policies, inflightPolicyExecutionIndices))
+            .map(IndexMetadata::getIndex)
+            .map(Index::getName)
+            .toArray(String[]::new);
+        deleteIndices(removeIndices);
     }
 
-    private boolean shouldRemoveIndex(GetIndexResponse getIndexResponse, Map<String, EnrichPolicy> policies, String indexName) {
-        // Find the policy on the index
+    private boolean shouldRemoveIndex(IndexMetadata indexMetadata, Map<String, EnrichPolicy> policies, Set<String> inflightPolicyIndices) {
+        String indexName = indexMetadata.getIndex().getName();
         logger.debug("Checking if should remove enrich index [{}]", indexName);
-        MappingMetadata mappingMetadata = getIndexResponse.getMappings().get(indexName);
+        // First ignore the index entirely if it is in the inflightPolicyIndices set as it is actively being worked on
+        if (inflightPolicyIndices.contains(indexName)) {
+            logger.debug("Enrich index [{}] was spared since it is reserved for an active policy execution.", indexName);
+            return false;
```

Comment:
Super minor nit, but this is very awkward for naming, because we have "should remove index" returning "false", but "remove" here means double the things (remove it from the list since it's called from a `filter` clause, or remove the index aka delete it).

Perhaps we could rename this function something like "indexUsedByPolicy" or something, and do a `== false` check on it in the filter clause? What do you think?